v2/spiffetls/dial_test.go

```diff
@@ -0,0 +1,301 @@
+			listenMode:       spiffetls.MTLSWebServerWithSource(tlsconfig.AuthorizeID(clientID), webCert, wlAPISourceA),
+			defaultWlAPIAddr: wlAPIServerB.Addr(),
+		},
+
+		// *WithSource Scenario
+		{
+			name:       "TLSClientWithSource succeeds",
+			dialMode:   spiffetls.TLSClientWithSource(tlsconfig.AuthorizeID(serverID), wlAPISourceB),
+			listenMode: spiffetls.TLSServerWithSource(wlAPISourceA),
+		},
+		{
+			name:       "MTLSClientWithSource succeeds",
+			dialMode:   spiffetls.MTLSClientWithSource(tlsconfig.AuthorizeID(serverID), wlAPISourceB),
+			listenMode: spiffetls.MTLSServerWithSource(tlsconfig.AuthorizeID(clientID), wlAPISourceA),
+		},
+		{
+			name:       "MTLSWebClient  succeeds",
+			dialMode:   spiffetls.MTLSWebClientWithSource(webCertPool, wlAPISourceB),
+			listenMode: spiffetls.MTLSWebServerWithSource(tlsconfig.AuthorizeID(clientID), webCert, wlAPISourceA),
+		},
+
+		// *WithSourceOptions Scenario
+		{
+			name:       "TLSClientWithSource succeeds",
+			dialMode:   spiffetls.TLSClientWithSourceOptions(tlsconfig.AuthorizeID(serverID), workloadapi.WithClient(wlAPIClientB)),
+			listenMode: spiffetls.TLSServerWithSource(wlAPISourceA),
+		},
+		{
+			name:       "MTLSClientWithSource succeeds",
+			dialMode:   spiffetls.MTLSClientWithSourceOptions(tlsconfig.AuthorizeID(serverID), workloadapi.WithClient(wlAPIClientB)),
+			listenMode: spiffetls.MTLSServerWithSource(tlsconfig.AuthorizeID(clientID), wlAPISourceA),
+		},
+		{
+			name:       "MTLSWebClient  succeeds",
+			dialMode:   spiffetls.MTLSWebClientWithSourceOptions(webCertPool, workloadapi.WithClient(wlAPIClientB)),
+			listenMode: spiffetls.MTLSWebServerWithSource(tlsconfig.AuthorizeID(clientID), webCert, wlAPISourceA),
+		},
+
+		// *WithRawConfig Scenario
+		{
+			name:       "TLSClientWithSource succeeds",
+			dialMode:   spiffetls.TLSClientWithRawConfig(tlsconfig.AuthorizeID(serverID), bundleSource),
+			listenMode: spiffetls.TLSServerWithSource(wlAPISourceA),
+		},
+		{
+			name:       "MTLSClientWithSource succeeds",
+			dialMode:   spiffetls.MTLSClientWithRawConfig(tlsconfig.AuthorizeID(serverID), svidSource, bundleSource),
+			listenMode: spiffetls.MTLSServerWithSource(tlsconfig.AuthorizeID(clientID), wlAPISourceA),
+		},
+		{
+			name:       "MTLSWebClient  succeeds",
+			dialMode:   spiffetls.MTLSWebClientWithRawConfig(webCertPool, svidSource),
+			listenMode: spiffetls.MTLSWebServerWithSource(tlsconfig.AuthorizeID(clientID), webCert, wlAPISourceA),
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+
+		err := os.Setenv("SPIFFE_ENDPOINT_SOCKET", test.defaultWlAPIAddr)
+		require.NoError(t, err)
+
+		t.Run(test.name, func(t *testing.T) {
+			// Start listening
+			listenCtx, cancelListenCtx := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancelListenCtx()
+
+			var listener net.Listener
+			var listenAddr string
+			listenDoneCh := make(chan struct{})
+			listenDataCh := make(chan string)
+			listenErrCh := make(chan error)
```

Comment:
These channels should be buffered. Otherwise, if the select below ends up failing the test, these goroutines will still be able to write into the channel. Otherwise, they will hang since there will be no receiving end.
```suggestion
			listenDoneCh := make(chan struct{}, 1)
			listenDataCh := make(chan string, 1)
			listenErrCh := make(chan error, 1)
```